File: src/hexa_display/hexa_display/display_node.py
"""Display relay node.

Subscribes to the gait engine state (`/gait/state`), the latest body
velocity command (`/cmd_vel`), the user body pose (`/body/pose`), the
posture animation mode (`/animation/mode`), and the battery state.
On a fixed timer it runs the pure expression policy and relays
SET_EXPRESSION / SET_GAZE frames to the ESP32 face over the configured
transport (`serial` on the robot, `stub` in sim — the stub logs the
decoded frames instead).

Pure sink: nothing in the workspace subscribes to or imports this
package. Fire-and-forget TX — the firmware animates autonomously and
NACKs are only logged. The transport is retried in the background, so
the robot comes up (and stays up) faceless if the display is absent.

Heartbeat: the firmware counts any valid frame as link activity and
falls back to a DEAD face after 3 s of silence, so whenever nothing
else has been written for ``ping_period_s`` (idle or steady-state
walking, where change detection sends nothing between refreshes) the
node sends a PING. The firmware echoes each PING back as a PONG; if
none arrives within ``pong_timeout_s`` the node logs a ROS error
(throttled) until the display responds again.

Face animations: while the policy selects one (breathing during
stack bringup, idling once the robot stands idle), this node runs its
clock and relays the due gaze/blink steps; the animation owns the
gaze until it ends.
"""

import rclpy
from geometry_msgs.msg import Twist
from hexa_interfaces.msg import BodyPose as BodyPoseMsg
from rclpy.node import Node
from rclpy.qos import DurabilityPolicy, QoSProfile, qos_profile_sensor_data
from sensor_msgs.msg import BatteryState
from std_msgs.msg import String

from .expression_policy import (
    DEFAULT_EXPRESSION_MAP,
    IDLE_TARGET,
    BatteryMonitor,
    DisplayTarget,
    PolicyConfig,
    PolicyInputs,
    decide,
    select_face_animation,
)
from .face_animation import FACE_ANIMATIONS, IDLING, FaceAnimation, due_steps
from .protocol import Cmd, Expression, NackReason, decode_frames
from .protocol import ping as ping_frame
from .protocol import set_expression as set_expression_frame
from .protocol import set_gaze as set_gaze_frame
from .protocol import trigger_blink as trigger_blink_frame
from .transport import (
    SerialTransport,
    StubTransport,
    Transport,
    TransportError,
)

WARN_THROTTLE_S = 10.0


def _parse_expression(name: str, param: str) -> Expression:
    key = name.strip().upper()
    if key not in Expression.__members__:
        valid = ", ".join(m.lower() for m in Expression.__members__)
        raise ValueError(f"{param}: unknown expression {name!r}; valid: {valid}")
    return Expression[key]


class DisplayNode(Node):
    def __init__(self) -> None:
        super().__init__("display_node")

        self._gait_state: str | None = None
        self._cmd_vel = Twist()
        self._body_pose = BodyPoseMsg()
        self._animation_mode: str = ""
        self._battery_voltage: float | None = None
        self._last_target: DisplayTarget = IDLE_TARGET
        self._sent_expression: Expression | None = None
        self._sent_gaze = None
        self._rx_buf = b""
        self._last_tx_t: float | None = None
        self._awaiting_pong_since: float | None = None
        self._pong_lost = False
        self._last_refresh_t: float | None = None
        self._last_reconnect_t: float | None = None
        self._active_face_animation: FaceAnimation | None = None
        self._face_animation_start_t = 0.0
        self._face_animation_fired = 0
        self._pending_face_animation: str | None = None
        self._pending_face_animation_since = 0.0

        self.declare_parameter("transport", "serial")
        self.declare_parameter("serial_device", "/dev/serial0")
        self.declare_parameter("serial_baud", 921600)
        self.declare_parameter("reconnect_period_s", 2.0)
        self.declare_parameter("update_rate_hz", 10.0)
        self.declare_parameter("refresh_period_s", 5.0)
        self.declare_parameter("ping_period_s", 1.0)
        self.declare_parameter("pong_timeout_s", 1.0)
        for state, expression in DEFAULT_EXPRESSION_MAP.items():
            self.declare_parameter(
                f"expression_map.{state}", expression.name.lower()
            )
        self.declare_parameter("animation_expression", "woozy")
        self.declare_parameter(
            "battery_topic", "/hexa_hardware_aux/battery_state"
        )
        self.declare_parameter("battery_warning_expression", "sleepy")
        self.declare_parameter("battery_critical_expression", "dead")
        self.declare_parameter("battery_warning_v", 0.0)
        self.declare_parameter("battery_critical_v", 0.0)
        self.declare_parameter("battery_hysteresis_v", 0.3)
        self.declare_parameter("battery_hold_s", 3.0)
        self.declare_parameter("gaze_deadband", 0.15)
        self.declare_parameter("gaze_exit_ratio", 0.6)
        self.declare_parameter("gaze_wz_weight", 1.0)
        self.declare_parameter("gaze_vy_max", 0.1)
        self.declare_parameter("gaze_wz_max", 0.5)
        self.declare_parameter("pose_pitch_threshold_rad", 0.08)
        self.declare_parameter("pose_tilt_threshold_rad", 0.08)
        self.declare_parameter("idling_start_delay_s", 4.0)

        def _str(name: str) -> str:
            return self.get_parameter(name).get_parameter_value().string_value

        def _dbl(name: str) -> float:
            return self.get_parameter(name).get_parameter_value().double_value

        # Fail fast on expression-name typos in the YAML.
        expression_map = {
            state: _parse_expression(
                _str(f"expression_map.{state}"), f"expression_map.{state}"
            )
            for state in DEFAULT_EXPRESSION_MAP
        }
        self._config = PolicyConfig(
            expression_map=expression_map,
            animation_expression=_parse_expression(
                _str("animation_expression"), "animation_expression"
            ),
            battery_warning_expression=_parse_expression(
                _str("battery_warning_expression"), "battery_warning_expression"
            ),
            battery_critical_expression=_parse_expression(
                _str("battery_critical_expression"), "battery_critical_expression"
            ),
            gaze_deadband=_dbl("gaze_deadband"),
            gaze_exit_ratio=_dbl("gaze_exit_ratio"),
            gaze_wz_weight=_dbl("gaze_wz_weight"),
            gaze_vy_max=_dbl("gaze_vy_max"),
            gaze_wz_max=_dbl("gaze_wz_max"),
            pose_pitch_threshold_rad=_dbl("pose_pitch_threshold_rad"),
            pose_tilt_threshold_rad=_dbl("pose_tilt_threshold_rad"),
            idling_start_delay_s=_dbl("idling_start_delay_s"),
        )
        self._battery_monitor = BatteryMonitor(
            warning_v=_dbl("battery_warning_v"),
            critical_v=_dbl("battery_critical_v"),
            hysteresis_v=_dbl("battery_hysteresis_v"),
            hold_s=_dbl("battery_hold_s"),
        )
        self._reconnect_period_s = _dbl("reconnect_period_s")
        self._refresh_period_s = _dbl("refresh_period_s")
        self._ping_period_s = _dbl("ping_period_s")
        self._pong_timeout_s = _dbl("pong_timeout_s")

        self._transport = self._make_transport(_str("transport"))
        try:
            self._transport.open()
        except TransportError as e:
            # The robot must come up faceless: keep retrying from the
            # tick timer instead of crashing.
            self.get_logger().warn(f"display transport unavailable: {e}")

        self._sub_gait_state = self.create_subscription(
            String, "/gait/state", self._on_gait_state, 10
        )
        self._sub_vel = self.create_subscription(
            Twist, "/cmd_vel", self._on_vel, 10
        )
        self._sub_pose = self.create_subscription(
            BodyPoseMsg, "/body/pose", self._on_pose, 10
        )
        # transient_local to match the teleop publisher, so a late
        # display start still sees the active animation mode.
        animation_qos = QoSProfile(
            depth=1, durability=DurabilityPolicy.TRANSIENT_LOCAL
        )
        self._sub_animation_mode = self.create_subscription(
            String, "/animation/mode", self._on_animation_mode, animation_qos
        )
        self._sub_battery = self.create_subscription(
            BatteryState,
            _str("battery_topic"),
            self._on_battery,
            qos_profile_sensor_data,
        )

        rate = _dbl("update_rate_hz")
        self._timer = self.create_timer(1.0 / rate, self._tick)

    def _make_transport(self, kind: str) -> Transport:
        if kind == "serial":
            device = (
                self.get_parameter("serial_device")
                .get_parameter_value()
                .string_value
            )
            baud = (
                self.get_parameter("serial_baud")
                .get_parameter_value()
                .integer_value
            )
            self.get_logger().info(f"display transport: serial {device} @ {baud}")
            return SerialTransport(device, baud)
        if kind == "stub":
            self.get_logger().info("display transport: stub (logging frames)")
            return StubTransport(
                log_fn=lambda msg: self.get_logger().info(f"display: {msg}")
            )
        raise ValueError(f"transport: unknown kind {kind!r} (serial | stub)")

    def _on_gait_state(self, msg: String) -> None:
        self._gait_state = msg.data

    def _on_vel(self, msg: Twist) -> None:
        self._cmd_vel = msg

    def _on_pose(self, msg: BodyPoseMsg) -> None:
        self._body_pose = msg

    def _on_animation_mode(self, msg: String) -> None:
        self._animation_mode = msg.data

    def _on_battery(self, msg: BatteryState) -> None:
        self._battery_voltage = msg.voltage

    def _now_s(self) -> float:
        return self.get_clock().now().nanoseconds * 1e-9

    def _ensure_transport(self, now: float) -> bool:
        if self._transport.is_open:
            return True
        if (
            self._last_reconnect_t is not None
            and now - self._last_reconnect_t < self._reconnect_period_s
        ):
            return False
        self._last_reconnect_t = now
        try:
            self._transport.open()
        except TransportError as e:
            self.get_logger().warn(
                f"display transport reconnect failed: {e}",
                throttle_duration_sec=WARN_THROTTLE_S,
            )
            return False
        self.get_logger().info("display transport reconnected")
        # Push full state on reconnect; pending PONGs from the old
        # link are void. _pong_lost survives so recovery is logged.
        self._sent_expression = None
        self._sent_gaze = None
        self._awaiting_pong_since = None
        return True

    def _write_frame(self, frame: bytes, now: float) -> bool:
        try:
            self._transport.write(frame)
            self._last_tx_t = now
            return True
        except TransportError as e:
            self.get_logger().warn(
                f"display write failed: {e}",
                throttle_duration_sec=WARN_THROTTLE_S,
            )
            self._sent_expression = None
            self._sent_gaze = None
            return False

    def _send_target(
        self, target: DisplayTarget, now: float, suppress_gaze: bool
    ) -> None:
        refresh = (
            self._last_refresh_t is None
            or now - self._last_refresh_t >= self._refresh_period_s
        )
        if refresh:
            self._last_refresh_t = now
        if refresh or target.expression != self._sent_expression:
            if self._write_frame(set_expression_frame(target.expression), now):
                self._sent_expression = target.expression
        if suppress_gaze:
            # A face animation owns the gaze; its own steps resync a
            # rebooted face within one cycle.
            return
        if refresh or target.gaze != self._sent_gaze:
            if self._write_frame(set_gaze_frame(target.gaze), now):
                self._sent_gaze = target.gaze

    def _update_face_animation(
        self, name: str | None, now: float
    ) -> FaceAnimation | None:
        if name is None:
            self._pending_face_animation = None
            self._active_face_animation = None
            return None
        if (
            self._active_face_animation is not None
            and self._active_face_animation.name == name
        ):
            return self._active_face_animation
        if self._pending_face_animation != name:
            self._pending_face_animation = name
            self._pending_face_animation_since = now
        delay = (
            self._config.idling_start_delay_s if name == IDLING.name else 0.0
        )
        if now - self._pending_face_animation_since < delay:
            self._active_face_animation = None
            return None
        self._active_face_animation = FACE_ANIMATIONS[name]
        self._face_animation_start_t = now
        self._face_animation_fired = 0
        return self._active_face_animation

    def _run_face_animation(self, animation: FaceAnimation, now: float) -> None:
        steps, self._face_animation_fired = due_steps(
            animation,
            now - self._face_animation_start_t,
            self._face_animation_fired,
        )
        for step in steps:
            if step.blink:
                self._write_frame(trigger_blink_frame(), now)
            if step.gaze is not None:
                if self._write_frame(set_gaze_frame(step.gaze), now):
                    self._sent_gaze = step.gaze

    def _drain_rx(self) -> None:
        try:
            data = self._transport.read()
        except TransportError:
            return  # write path already logs and schedules reconnect
        if not data:
            return
        frames, self._rx_buf = decode_frames(self._rx_buf + data)
        for frame in frames:
            if frame.cmd == Cmd.NACK:
                reason = "?"
                if len(frame.payload) == 1:
                    try:
                        reason = NackReason(frame.payload[0]).name
                    except ValueError:
                        reason = f"0x{frame.payload[0]:02X}"
                self.get_logger().warn(f"display NACK: {reason}")
            elif frame.cmd == Cmd.PONG:
                self._awaiting_pong_since = None
                if self._pong_lost:
                    self._pong_lost = False
                    self.get_logger().info("display responding again (PONG)")
            elif frame.cmd == Cmd.LOG:
                text = frame.payload.decode("utf-8", errors="replace")
                self.get_logger().warn(f"display log: {text}")
            else:
                self.get_logger().debug(
                    f"display rx: cmd=0x{frame.cmd:02X} "
                    f"payload={frame.payload.hex()}"
                )

    def _tick(self) -> None:
        now = self._now_s()
        battery_low, battery_critical = False, False
        if self._battery_voltage is not None:
            battery_low, battery_critical = self._battery_monitor.update(
                self._battery_voltage, now
            )
        inputs = PolicyInputs(
            gait_state=self._gait_state,
            vx=self._cmd_vel.linear.x,
            vy=self._cmd_vel.linear.y,
            wz=self._cmd_vel.angular.z,
            animation_mode=self._animation_mode,
            roll=self._body_pose.roll,
            pitch=self._body_pose.pitch,
            yaw=self._body_pose.yaw,
            battery_low=battery_low,
            battery_critical=battery_critical,
        )
        self._last_target = decide(inputs, self._config, self._last_target)
        animation = self._update_face_animation(
            select_face_animation(inputs, self._config), now
        )
        if not self._ensure_transport(now):
            return
        self._send_target(
            self._last_target, now, suppress_gaze=animation is not None
        )
        if animation is not None:
            self._run_face_animation(animation, now)
        # Heartbeat: the firmware drops the link (DEAD face) after 3 s
        # of silence, so keep traffic flowing when change detection has
        # nothing to send — PING counts as link activity. Each PING is
        # echoed back as a PONG; _awaiting_pong_since tracks the oldest
        # unanswered one.
        if (
            self._last_tx_t is None
            or now - self._last_tx_t >= self._ping_period_s
        ):
            if (
                self._write_frame(ping_frame(), now)
                and self._awaiting_pong_since is None
            ):
                self._awaiting_pong_since = now
        self._drain_rx()
        if (
            self._awaiting_pong_since is not None
            and now - self._awaiting_pong_since >= self._pong_timeout_s
        ):
            self._pong_lost = True
            self.get_logger().error(
                "display not responding: no PONG within "
                f"{self._pong_timeout_s:.1f} s",
                throttle_duration_sec=WARN_THROTTLE_S,
            )


def main(args=None) -> None:
    rclpy.init(args=args)
    node = DisplayNode()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/hexa_display/src/display_node.cpp
// Hexapod "face" ROS2 node. One process maps robot state to an expression/gaze
// policy and rasterizes the eyes on a 256x64 SH1122 OLED directly from the Pi,
// reusing the firmware's platform-free eye animation core (EyeAnim + EyeRaster).
//
// A pure sink of robot state: it subscribes to the gait engine state, the body
// velocity command, the user body pose, the posture animation mode, the
// battery state, and the Bluetooth pairing state, and drives the renderer
// target directly — the policy half and the render half share one process, so
// there is no /display/* topic hop.
//
// Text mode: a non-empty /display/text message replaces the eyes with the
// message (Pixel Operator font, word-wrapped, centered); an empty message
// returns to the face. The policy keeps ticking underneath so the face resumes
// with current state.

#include <chrono>
#include <cstdint>
#include <memory>
#include <optional>
#include <random>
#include <stdexcept>
#include <string>

#include <geometry_msgs/msg/twist.hpp>
#include <hexa_interfaces/msg/body_pose.hpp>
#include <rcl_interfaces/msg/parameter_descriptor.hpp>
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/battery_state.hpp>
#include <std_msgs/msg/bool.hpp>
#include <std_msgs/msg/string.hpp>

#include "EyeAnim.h"
#include "EyeRaster.h"
#include "Expression.h"
#include "FaceNames.h"
#include "Sh1122Panel.h"
#include "u8g2.h"

#include "expression_policy.hpp"
#include "face_animation.hpp"
#include "face_animation_runner.hpp"
#include "text_screen.hpp"

namespace {

// EyeAnim's injected RNG. A process-wide mt19937, seeded once, satisfies the
// uint32_t(*)() signature (a plain function pointer cannot capture state).
uint32_t hostRand() {
    static std::mt19937 gen{std::random_device{}()};
    return gen();
}

void pixelSink(void* ctx, int x, int y) {
    u8g2_DrawPixel(static_cast<u8g2_t*>(ctx),
                   static_cast<u8g2_uint_t>(x), static_cast<u8g2_uint_t>(y));
}

// Exact equality is intended: drawEye is a pure function of these fields, so
// bit-identical frames rasterize to identical pixels. EyeAnim emits a constant
// frame while idle (lid == 1.0f, settled integer gaze, phase 0), so this
// reliably fires. SCANNING steps its phase in discrete jumps, so the spinner
// redraws once per step rather than once per render tick.
bool sameFrame(const eyes::AnimFrame& a, const eyes::AnimFrame& b) {
    return a.expr == b.expr && a.lid == b.lid && a.gx == b.gx && a.gy == b.gy &&
           a.phase == b.phase;
}

rcl_interfaces::msg::ParameterDescriptor readOnly(const std::string& desc) {
    rcl_interfaces::msg::ParameterDescriptor d;
    d.description = desc;
    d.read_only = true;
    return d;
}

std::string toLower(std::string s) {
    for (char& c : s) c = static_cast<char>(std::tolower(static_cast<unsigned char>(c)));
    return s;
}

}  // namespace

using hexa::display::BatteryMonitor;
using hexa::display::decide;
using hexa::display::DisplayTarget;
using hexa::display::FaceAnimation;
using hexa::display::FaceAnimationRunner;
using hexa::display::PolicyConfig;
using hexa::display::PolicyInputs;
using hexa::display::selectFaceAnimation;

class DisplayNode : public rclcpp::Node {
public:
    DisplayNode()
        : Node("display_node"),
          _anim(&hostRand),
          _runner(
              [this](double lo, double hi) {
                  std::uniform_real_distribution<double> dist(lo, hi);
                  return dist(_rng);
              },
              [this](GazeDirection g) { _target.gaze = g; },
              [this]() { _anim.requestBlink(); }) {
        _rng.seed(std::random_device{}());

        // --- Panel / render params (read-only, as the renderer half) ---
        face::PanelConfig cfg;
        cfg.spi_device = declare_parameter<std::string>(
            "spi_device", cfg.spi_device, readOnly("spidev path for the SH1122"));
        cfg.spi_speed_hz = static_cast<uint32_t>(declare_parameter<int>(
            "spi_speed_hz", static_cast<int>(cfg.spi_speed_hz),
            readOnly("SPI clock (Hz)")));
        cfg.gpio_chip = declare_parameter<std::string>(
            "gpio_chip", cfg.gpio_chip, readOnly("GPIO character device"));
        cfg.dc_line = declare_parameter<int>("dc_line", 24, readOnly("DC GPIO line"));
        cfg.rst_line = declare_parameter<int>("rst_line", 25, readOnly("RST GPIO line"));
        cfg.cs_line = declare_parameter<int>(
            "cs_line", -1, readOnly("CS GPIO line, -1 to leave CS to the controller"));
        cfg.headless = declare_parameter<bool>(
            "headless", false, readOnly("run without touching SPI/GPIO"));
        const int render_hz = std::max(1, static_cast<int>(declare_parameter<int>(
            "render_hz", 60, readOnly("render loop rate (Hz)"))));

        // --- Policy params ---
        const double update_rate_hz = declare_parameter<double>("update_rate_hz", 10.0);
        for (const auto& [state, expression] : hexa::display::defaultExpressionMap()) {
            declare_parameter<std::string>("expression_map." + state,
                                           toLower(expressionName(expression)));
        }
        declare_parameter<std::string>("animation_expression", "woozy");
        const std::string battery_topic = declare_parameter<std::string>(
            "battery_topic", "/hexa_hardware_aux/battery_state");
        declare_parameter<std::string>("battery_warning_expression", "sleepy");
        declare_parameter<std::string>("battery_critical_expression", "dead");
        declare_parameter<std::string>("scanning_expression", "scanning");
        declare_parameter<double>("battery_warning_v", 0.0);
        declare_parameter<double>("battery_critical_v", 0.0);
        declare_parameter<double>("battery_hysteresis_v", 0.3);
        declare_parameter<double>("battery_hold_s", 3.0);
        declare_parameter<double>("gaze_deadband", 0.15);
        declare_parameter<double>("gaze_exit_ratio", 0.6);
        declare_parameter<double>("gaze_wz_weight", 1.0);
        declare_parameter<double>("gaze_vy_max", 0.1);
        declare_parameter<double>("gaze_wz_max", 0.5);
        declare_parameter<double>("pose_pitch_threshold_rad", 0.08);
        declare_parameter<double>("pose_tilt_threshold_rad", 0.08);
        declare_parameter<double>("idling_start_delay_s", 4.0);

        // Fail fast on expression-name typos in the YAML.
        for (const auto& [state, expression] : hexa::display::defaultExpressionMap()) {
            (void)expression;
            _config.expression_map[state] = parseExpressionParam("expression_map." + state);
        }
        _config.animation_expression = parseExpressionParam("animation_expression");
        _config.battery_warning_expression = parseExpressionParam("battery_warning_expression");
        _config.battery_critical_expression = parseExpressionParam("battery_critical_expression");
        _config.scanning_expression = parseExpressionParam("scanning_expression");
        _config.gaze_deadband = get_parameter("gaze_deadband").as_double();
        _config.gaze_exit_ratio = get_parameter("gaze_exit_ratio").as_double();
        _config.gaze_wz_weight = get_parameter("gaze_wz_weight").as_double();
        _config.gaze_vy_max = get_parameter("gaze_vy_max").as_double();
        _config.gaze_wz_max = get_parameter("gaze_wz_max").as_double();
        _config.pose_pitch_threshold_rad = get_parameter("pose_pitch_threshold_rad").as_double();
        _config.pose_tilt_threshold_rad = get_parameter("pose_tilt_threshold_rad").as_double();
        _config.idling_start_delay_s = get_parameter("idling_start_delay_s").as_double();

        _battery = BatteryMonitor(
            get_parameter("battery_warning_v").as_double(),
            get_parameter("battery_critical_v").as_double(),
            get_parameter("battery_hysteresis_v").as_double(),
            get_parameter("battery_hold_s").as_double());

        if (!_panel.begin(cfg)) {
            RCLCPP_FATAL(get_logger(),
                         "SH1122 init failed (spi=%s chip=%s). Check wiring, "
                         "dtparam=spi=on, and access to spidev/gpiochip "
                         "(groups spi/gpio), or set headless:=true.",
                         cfg.spi_device.c_str(), cfg.gpio_chip.c_str());
            throw std::runtime_error("panel init failed");
        }

        // --- Inputs: cache the latest of each; the policy tick snapshots them ---
        // transient_local to match hexa_locomotion's latched publisher: the boot
        // "folded" goes out before this node subscribes.
        rclcpp::QoS gait_qos(1);
        gait_qos.transient_local();
        _sub_gait = create_subscription<std_msgs::msg::String>(
            "/gait/state", gait_qos,
            [this](const std_msgs::msg::String& m) { _gait_state = m.data; });
        _sub_vel = create_subscription<geometry_msgs::msg::Twist>(
            "/cmd_vel", 10,
            [this](const geometry_msgs::msg::Twist& m) { _cmd_vel = m; });
        _sub_pose = create_subscription<hexa_interfaces::msg::BodyPose>(
            "/body/pose", 10,
            [this](const hexa_interfaces::msg::BodyPose& m) { _body_pose = m; });
        // transient_local to match the teleop publisher, so a late display start
        // still sees the active animation mode.
        rclcpp::QoS animation_qos(1);
        animation_qos.transient_local();
        _sub_animation = create_subscription<std_msgs::msg::String>(
            "/animation/mode", animation_qos,
            [this](const std_msgs::msg::String& m) { _animation_mode = m.data; });
        _sub_battery = create_subscription<sensor_msgs::msg::BatteryState>(
            battery_topic, rclcpp::SensorDataQoS(),
            [this](const sensor_msgs::msg::BatteryState& m) { _battery_voltage = m.voltage; });
        // Text mode input. transient_local so a late display start still shows
        // the message; an empty string hands the panel back to the face.
        rclcpp::QoS text_qos(1);
        text_qos.transient_local();
        _sub_text = create_subscription<std_msgs::msg::String>(
            "/display/text", text_qos,
            [this](const std_msgs::msg::String& m) { _display_text = m.data; });
        // Pairing state from hexa_buttons. transient_local so a display restart
        // mid-scan comes back up wearing the spinners rather than the face.
        rclcpp::QoS scanning_qos(1);
        scanning_qos.transient_local();
        _sub_scanning = create_subscription<std_msgs::msg::Bool>(
            "/bluetooth/scanning", scanning_qos,
            [this](const std_msgs::msg::Bool& m) { _bluetooth_scanning = m.data; });
        // The other thing a front-panel hold starts: a network-mode switch.
        // Separate topic because it means something else, same spinners because
        // the face's answer to "wait, I am working" is the same either way.
        _sub_busy = create_subscription<std_msgs::msg::Bool>(
            "/display/busy", scanning_qos,
            [this](const std_msgs::msg::Bool& m) { _display_busy = m.data; });

        // Policy tick on the node clock (sim-time aware). The renderer eases gaze
        // and blinks autonomously, so a low rate suffices.
        _policy_timer = create_wall_timer(
            std::chrono::duration_cast<std::chrono::nanoseconds>(
                std::chrono::duration<double>(1.0 / update_rate_hz)),
            [this] { policyTick(); });
        // Render loop: wall-clock, decoupled from the policy rate. Frames only
        // reach SPI when the pixels change.
        _render_timer = create_wall_timer(
            std::chrono::milliseconds(1000 / render_hz), [this] { renderTick(); });

        RCLCPP_INFO(get_logger(), "face up: 256x64 SH1122 on %s @ %d Hz%s",
                    cfg.spi_device.c_str(), render_hz,
                    cfg.headless ? " (headless)" : "");
    }

    ~DisplayNode() override { _panel.sleep(); }

private:
    Expression parseExpressionParam(const std::string& param) {
        const std::string value = get_parameter(param).as_string();
        if (auto e = face::parseExpression(value)) {
            return *e;
        }
        std::string valid;
        for (int i = 0; i < static_cast<int>(Expression::_COUNT); ++i) {
            if (i) valid += ", ";
            valid += toLower(expressionName(static_cast<Expression>(i)));
        }
        throw std::runtime_error(param + ": unknown expression '" + value +
                                 "'; valid: " + valid);
    }

    uint32_t nowMs() {
        using namespace std::chrono;
        if (_t0.time_since_epoch().count() == 0) _t0 = steady_clock::now();
        return static_cast<uint32_t>(
            duration_cast<milliseconds>(steady_clock::now() - _t0).count());
    }

    void policyTick() {
        const double now = get_clock()->now().seconds();
        BatteryMonitor::Flags battery;
        if (_battery_voltage) {
            battery = _battery.update(*_battery_voltage, now);
        }
        PolicyInputs inputs;
        inputs.gait_state = _gait_state;
        inputs.vx = _cmd_vel.linear.x;
        inputs.vy = _cmd_vel.linear.y;
        inputs.wz = _cmd_vel.angular.z;
        inputs.animation_mode = _animation_mode;
        inputs.roll = _body_pose.roll;
        inputs.pitch = _body_pose.pitch;
        inputs.yaw = _body_pose.yaw;
        inputs.battery_low = battery.low;
        inputs.battery_critical = battery.critical;
        inputs.busy = _bluetooth_scanning || _display_busy;

        _last_target = decide(inputs, _config, _last_target);
        const FaceAnimation* animation = _runner.update(
            selectFaceAnimation(inputs, _config), now, _config.idling_start_delay_s);

        // Expression always comes from the policy. Gaze too, unless a face
        // animation is active — then its own steps own the gaze (and its
        // gaze_ease_s the drift speed).
        _target.expr = _last_target.expression;
        if (animation == nullptr) {
            _target.gaze = _last_target.gaze;
            _target.gazeEaseMs = 0;
            _target.gazeScale = 1.0f;
        } else {
            _runner.run(*animation, now);
            _target.gazeEaseMs =
                static_cast<uint32_t>(animation->gaze_ease_s() * 1000.0);
            _target.gazeScale = static_cast<float>(animation->gaze_scale());
        }
    }

    void renderTick() {
        if (!_display_text.empty()) {
            // Text mode: the message is static, so redraw only on change (the
            // string compare replaces the eye path's sameFrame skip).
            if (_display_text != _drawn_text) {
                _drawn_text = _display_text;
                _panel.clearBuffer();
                hexa::display::drawTextScreen(_panel.u8g2(), _display_text, _text_cfg);
                _panel.present();
            }
            _haveFrame = false;  // repaint the face when text mode ends
            return;
        }
        _drawn_text.clear();

        const eyes::AnimFrame f = _anim.update(_target, nowMs());
        // Skip the float-heavy raster entirely when the animation frame is
        // unchanged — between blinks and once gaze has settled the face is
        // static, so there is nothing to redraw.
        if (_haveFrame && sameFrame(f, _lastFrame)) return;
        _lastFrame = f;
        _haveFrame = true;

        _panel.clearBuffer();
        u8g2_t* g = _panel.u8g2();
        eyes::drawEye(f.expr, false, eyes::kEyeLX, f.lid, f.gx, f.gy, f.phase,
                      pixelSink, g);
        eyes::drawEye(f.expr, true, eyes::kEyeRX, f.lid, f.gx, f.gy, f.phase,
                      pixelSink, g);
        _panel.present();  // flushes over SPI only if the pixels changed
    }

    // Render half.
    face::Sh1122Panel _panel;
    eyes::EyeAnim _anim;
    std::mt19937 _rng;
    RenderState _target{Expression::NEUTRAL, GazeDirection::CENTER};
    eyes::AnimFrame _lastFrame{};
    bool _haveFrame = false;
    std::chrono::steady_clock::time_point _t0{};
    hexa::display::TextScreenConfig _text_cfg;
    std::string _drawn_text;  // text currently on the panel ("" = face)

    // Policy half.
    PolicyConfig _config;
    BatteryMonitor _battery;
    FaceAnimationRunner _runner;
    DisplayTarget _last_target = hexa::display::kIdleTarget;

    // Cached inputs.
    std::optional<std::string> _gait_state;
    geometry_msgs::msg::Twist _cmd_vel;
    hexa_interfaces::msg::BodyPose _body_pose;
    std::string _animation_mode;
    std::optional<double> _battery_voltage;
    std::string _display_text;
    bool _bluetooth_scanning = false;
    bool _display_busy = false;

    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr _sub_gait;
    rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr _sub_vel;
    rclcpp::Subscription<hexa_interfaces::msg::BodyPose>::SharedPtr _sub_pose;
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr _sub_animation;
    rclcpp::Subscription<sensor_msgs::msg::BatteryState>::SharedPtr _sub_battery;
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr _sub_text;
    rclcpp::Subscription<std_msgs::msg::Bool>::SharedPtr _sub_scanning;
    rclcpp::Subscription<std_msgs::msg::Bool>::SharedPtr _sub_busy;
    rclcpp::TimerBase::SharedPtr _policy_timer;
    rclcpp::TimerBase::SharedPtr _render_timer;
};

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    int rc = 0;
    try {
        rclcpp::spin(std::make_shared<DisplayNode>());
    } catch (const std::exception& e) {
        RCLCPP_FATAL(rclcpp::get_logger("display_node"), "fatal: %s", e.what());
        rc = 1;
    }
    rclcpp::shutdown();
    return rc;
}
